Implement [Hashtbl.set] using [replace]

rgrinberg opened the pull request:
Stdune's hash tables do not support multiple bindings.

While I'm at it, should we rename all the names in hashtable that involve `add` into `set`? We don't support multiple bindings so our naming should reflect that.

thanks @giltho!

aalekseyev commented:
Wow, I never knew `Hashtbl` was a multimap! I'll have to rethink my life now.

diml commented on src/stdune/hashtbl.ml:
```diff
@@ -8,14 +8,18 @@ end) =
 struct
   include MoreLabels.Hashtbl.Make (H)
 
+  [@@@ocaml.warning "-32"]
```
Can you attach this to just `add`?

diml commented:
Regarding the naming, we definitely shouldn't have both in the API because it is confusing. Right now, our `add`  is the `set` of the `stdlib`. In Base, `add` is also the `set` of the stdlib. So I think either naming is fine as long as there is only one. I'd use `add` because that's what `Set` and `Map` use and they mean to replace any previous binding.

To make things more robust, I propose to remove the `include MoreLabels.Hashtbl.S` from `hashtbl_intf.ml` so that we control more precisely the API of `Hashbtl`.

aalekseyev commented:
I think `add` is a terrible name for replacement of previous binding. Neither base nor core keep that name and I think we shouldn't either. (or maybe I misunderstood what @diml said?)

rgrinberg commented:
> I think add is a terrible name for replacement of previous binding. Neither base nor core keep that name and I think we shouldn't either.

`Base.Map.add_exn` and `Base.Hashtbl.add_exn` both exist.

I agree that `set` is a better name however.

aalekseyev commented:
`add_exn` doesn't replace an existing binding though: it always adds a new entry if it returns, so that name seems reasonable to me.

rgrinberg commented:
> add_exn doesn't replace an existing binding though: it always adds a new entry if it returns, so that name seems reasonable to me.

Okay, this makes sense.

@diml I've removed the `include` of stdlib's hashtable interface and fixed the attribute. This PR should be ready.

diml commented:
Hmm, yh never-mind. I forgot that in some cases we want the `set` semantic, for a minute I thought we only needed one of the two versions.
